File: guides/managed-agents/cursor-cloud-agents/python/build_template.py
"""Build the Superserve template that Cursor pool workers boot from."""

from __future__ import annotations

import os
from pathlib import Path

import dotenv
from superserve import RunStep, Template, WorkdirStep

dotenv.load_dotenv(Path(__file__).with_name(".env"))

TEMPLATE_NAME = os.environ.get("CURSOR_WORKER_TEMPLATE", "cursor-worker")

STEPS = [
    RunStep(
        run=(
            "apt-get update && apt-get install -y --no-install-recommends "
            "ca-certificates curl git jq procps unzip && rm -rf /var/lib/apt/lists/*"
        )
    ),
    # Installs to /root/.local/bin/agent; the symlink puts it on every PATH.
    RunStep(run="curl -fsS https://cursor.com/install | HOME=/root bash"),
    RunStep(
        run="ln -sf /root/.local/bin/agent /usr/local/bin/agent && agent --version"
    ),
    RunStep(run="mkdir -p /workspace /var/lib/cursor-worker"),
    WorkdirStep(workdir="/workspace"),
]


def _on_log(ev) -> None:
    if ev.stream.value != "system":
        print(ev.text, end="", flush=True)


def main() -> int:
    # Template names are unique per team, so a rerun must pick up the existing
    # template: reuse a ready one, wait on an in-flight build, rebuild a failed one.
    existing = next((t for t in Template.list() if t.name == TEMPLATE_NAME), None)
    if existing is not None:
        status = existing.status.value
        template = Template.connect(existing.id)
        if status == "ready":
            print(
                f"template {TEMPLATE_NAME!r} already exists and is ready (id: {existing.id})"
            )
            return 0
        if status == "failed":
            print(
                f"template {TEMPLATE_NAME!r} has a failed build (id: {existing.id}), rebuilding..."
            )
            template.rebuild()
            # Reconnect so wait_until_ready() tracks the new build, not the failed one.
            template = Template.connect(existing.id)
        else:
            print(
                f"template {TEMPLATE_NAME!r} is {status} (id: {existing.id}), waiting..."
            )
    else:
        print(f"creating template {TEMPLATE_NAME!r}...")
        template = Template.create(
            name=TEMPLATE_NAME,
            from_="ubuntu:24.04",
            vcpu=2,
            memory_mib=2048,
            disk_mib=8192,
            steps=STEPS,
        )
        print(f"template created (id: {template.id}), waiting for build...")

    template.wait_until_ready(on_log=_on_log)

    print(f"\ntemplate {TEMPLATE_NAME!r} is ready (id: {template.id})")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: guides/managed-agents/cursor-cloud-agents/typescript/build-template.mjs
#!/usr/bin/env node
import "./env.mjs"
import { Template } from "@superserve/sdk"

const TEMPLATE_NAME = process.env.CURSOR_WORKER_TEMPLATE || "cursor-worker"

const STEPS = [
  {
    run:
      "apt-get update && apt-get install -y --no-install-recommends " +
      "ca-certificates curl git jq procps unzip && rm -rf /var/lib/apt/lists/*",
  },
  // Installs to /root/.local/bin/agent; the symlink puts it on every PATH.
  { run: "curl -fsS https://cursor.com/install | HOME=/root bash" },
  {
    run: "ln -sf /root/.local/bin/agent /usr/local/bin/agent && agent --version",
  },
  { run: "mkdir -p /workspace /var/lib/cursor-worker" },
  { workdir: "/workspace" },
]

const onLog = (ev) => {
  if (ev.stream !== "system") process.stdout.write(ev.text)
}

// Template names are unique per team, so a rerun must pick up the existing
// template: reuse a ready one, wait on an in-flight build, rebuild a failed one.
const existing = (await Template.list()).find((t) => t.name === TEMPLATE_NAME)
let template
if (existing) {
  template = await Template.connect(existing.id)
  if (existing.status === "ready") {
    console.log(
      `template '${TEMPLATE_NAME}' already exists and is ready (id: ${existing.id})`,
    )
    process.exit(0)
  }
  if (existing.status === "failed") {
    console.log(
      `template '${TEMPLATE_NAME}' has a failed build (id: ${existing.id}), rebuilding...`,
    )
    await template.rebuild()
    // Reconnect so waitUntilReady() tracks the new build, not the failed one.
    template = await Template.connect(existing.id)
  } else {
    console.log(
      `template '${TEMPLATE_NAME}' is ${existing.status} (id: ${existing.id}), waiting...`,
    )
  }
} else {
  console.log(`creating template '${TEMPLATE_NAME}'...`)
  template = await Template.create({
    name: TEMPLATE_NAME,
    from: "ubuntu:24.04",
    vcpu: 2,
    memoryMib: 2048,
    diskMib: 8192,
    steps: STEPS,
  })
  console.log(`template created (id: ${template.id}), waiting for build...`)
}

await template.waitUntilReady({ onLog })

console.log(`\ntemplate '${TEMPLATE_NAME}' is ready (id: ${template.id})`)
